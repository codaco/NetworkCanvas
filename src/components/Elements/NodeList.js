import React from 'react';
import PropTypes from 'prop-types';
import cx from 'classnames';
import { animation } from 'network-canvas-ui';
import StaggeredTransitionGroup from '../../utils/StaggeredTransitionGroup';
import { scrollable, droppable, draggable, selectable } from '../../behaviours';
import Node from './Node';

const EnhancedNode = draggable(selectable(Node));

/**
  * Renders a list of Node.
  */
const NodeList = ({
  nodes,
  label,
  selected,
  handleSelectNode,
  handleDropNode,
  draggableType,
  hover,
}) => {
  const classNames = cx(
    'node-list',
    { 'node-list--hover': hover },
  );

  return (
    <StaggeredTransitionGroup
      className={classNames}
      component="div"
      delay={animation.duration.fast * 0.2}
      duration={animation.duration.slow}
      start={animation.duration.slow + 3}
      transitionName="node-list--transition"
      transitionLeave={false}
    >
      { hover &&
        <Node key="placeholder" placeholder />
      }
      {
        nodes.map(node => (
          <span key={node.uid}>
            <EnhancedNode
              label={label(node)}
              selected={selected(node)}
              onSelected={() => handleSelectNode(node)}
              onDropped={hits => handleDropNode(hits, node)}
              draggableType={draggableType}
              {...node}
            />
          </span>
        ))
      }
    </StaggeredTransitionGroup>
  );
};

NodeList.propTypes = {
  nodes: PropTypes.array.isRequired,
  handleSelectNode: PropTypes.func,
  handleDropNode: PropTypes.func,
  label: PropTypes.func,
  selected: PropTypes.func,
  draggableType: PropTypes.string,
  hover: PropTypes.bool,
};

NodeList.defaultProps = {
  nodes: [],
  label: () => (''),
  selected: () => false,
  handleSelectNode: () => {},
  handleDropNode: () => {},
  draggableType: '',
  hover: false,
};

export default droppable(scrollable(NodeList));
