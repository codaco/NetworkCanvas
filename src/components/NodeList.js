import React, { Component } from 'react';
import { compose } from 'redux';
import PropTypes from 'prop-types';
import { find, get, isEqual } from 'lodash';
import cx from 'classnames';
import Color from 'color';
import { Node, colorDictionary } from 'network-canvas-ui';
import { TransitionGroup } from 'react-transition-group';
import { Node as NodeTransition } from './Transition';
import { scrollable, selectable } from '../behaviours';
import {
  DragSource,
  DropTarget,
  MonitorDropTarget,
  MonitorDragSource,
} from '../behaviours/DragAndDrop';

const EnhancedNode = DragSource(selectable(Node));

/**
 * @class NodeList
 *
 * Renders a list of Nodes.
 *
 * @param { string } props.listId A unique identifier for this list.
 *    Note: `key` is not used, to simplify the animations used to re-render lists.
 *
 * @description
 * Notes on state/props:
 *
 * `props.nodes` are copied to state so that they can be modified out of sync
 * from props. For example, so that we can modify the transition properties
 * of previously-displayed nodes in response to props changes before displaying
 * the new nodes.
 *
 * `listId` is similarly copied to state, as it is used (in conjunction with
 * node.uid) to identify a group of nodes. If the same node appears in two lists,
 * then it is logically separate so that it still disappears & reappears with
 * the rest of its group.
 */
class NodeList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      nodes: props.nodes,
      stagger: true,
      exit: true,
      waitForExitAnimation: true,
    };

    this.refreshTimer = null;
  }

  componentWillReceiveProps(newProps) {
    // Don't update node display if nodes are the same
    if (isEqual(newProps.nodes, this.state.nodes)) {
      return;
    }

    // if we provided the same id, then just update normally
    if (newProps.listId === this.props.listId) {
      this.setState({ exit: false }, () => {
        this.setState({
          nodes: newProps.nodes,
          waitForExitAnimation: false,
          stagger: false,
        });
      });
      return;
    }

    // Otherwise, transition out and in again
    this.props.scrollTop(0);

    // First, ensure that exit animations are enabled for previous nodes
    this.setState({ exit: true }, () => {
      // Then transition new ones in
      this.setState({
        listId: newProps.listId,
        nodes: newProps.nodes,
        stagger: true,
        // Even if there were no previous nodes, include the pause
        waitForExitAnimation: true,
      });
    });
  }

  render() {
    const {
      nodeColor,
      label,
      selected,
      onSelect,
      itemType,
      isOver,
      willAccept,
      meta,
    } = this.props;

    const {
      stagger,
      exit,
      listId,
      nodes,
      waitForExitAnimation,
    } = this.state;

    const isSource = !!find(nodes, ['uid', get(meta, 'uid', null)]);
    const isValidTarget = !isSource && willAccept;
    const isHovering = isValidTarget && isOver;

    const classNames = cx(
      'node-list',
      { 'node-list--drag': isValidTarget },
    );

    const backgroundColor = Color(
      nodeColor || colorDictionary['node-base'],
    ).alpha(0.1);

    const styles = isHovering ? { backgroundColor } : {};

    return (
      <TransitionGroup
        className={classNames}
        style={styles}
        exit={exit}
      >
        {
          nodes.map((node, index) => (
            <NodeTransition
              key={`${listId}-${node.uid}`}
              index={index}
              stagger={stagger}
              waitForPreviousExit={waitForExitAnimation}
              listId={listId}
            >
              <EnhancedNode
                color={nodeColor}
                label={`${label(node)}`}
                selected={selected(node)}
                onSelected={() => onSelect(node)}
                meta={() => ({ ...node, itemType })}
                {...node}
              />
            </NodeTransition>
          ))
        }
      </TransitionGroup>
    );
  }
}

NodeList.propTypes = {
  nodes: PropTypes.array.isRequired,
  nodeColor: PropTypes.string,
  onSelect: PropTypes.func,
  itemType: PropTypes.string,
  label: PropTypes.func,
  selected: PropTypes.func,
  isOver: PropTypes.bool,
  willAccept: PropTypes.bool,
  meta: PropTypes.object,
  listId: PropTypes.string.isRequired,
  scrollTop: PropTypes.func,
};

NodeList.defaultProps = {
  nodes: [],
  nodeColor: '',
  label: () => (''),
  selected: () => false,
  onSelect: () => {},
  onDrop: () => {},
  itemType: 'NODE',
  isOver: false,
  willAccept: false,
  isDragging: false,
  meta: {},
  scrollTop: () => {},
};

export default compose(
  DropTarget,
  MonitorDropTarget(['isOver', 'willAccept']),
  MonitorDragSource(['meta', 'isDragging']),
  scrollable,
)(NodeList);
