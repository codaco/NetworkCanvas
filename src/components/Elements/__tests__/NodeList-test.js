/* eslint-env jest */

import React from 'react';
import { createStore } from 'redux';
import { shallow } from 'enzyme';
import NodeList from '../../Elements/NodeList';

const network = {
  nodes: [
    {
      name: 'test',
    },
  ],
};

describe('NodeList component', () => {
  it('renders ok', () => {
    const component = shallow((
      <NodeList
        network={network}
        store={createStore(() => {})}
      />
    ));

    expect(component).toMatchSnapshot();
  });
});
