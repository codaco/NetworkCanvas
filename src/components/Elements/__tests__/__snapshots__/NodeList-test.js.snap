exports[`NodeList component renders ok 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <Droppable
    acceptsDraggableType={null}
    droppableName={null}
    network={
        Object {
            "nodes": Array [
              Object {
                "name": "test",
              },
            ],
          }
    }
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    }
    updateZone={[Function]} />,
  "nodes": Array [
    <Droppable
      acceptsDraggableType={null}
      droppableName={null}
      network={
            Object {
                "nodes": Array [
                  Object {
                    "name": "test",
                  },
                ],
              }
        }
        store={
            Object {
                "dispatch": [Function],
                "getState": [Function],
                "replaceReducer": [Function],
                "subscribe": [Function],
                Symbol(observable): [Function],
              }
        } />
</div>,
  "nodes": Array [
    <div
      className="scrollable">
      <Connect(Droppable)
            network={
                  Object {
                        "nodes": Array [
                          Object {
                            "name": "test",
                          },
                        ],
                      }
            }
            store={
                  Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
            } />
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 2,
      "_context": Object {},
      "_currentElement": <Scrollable
        network={
                Object {
                        "nodes": Array [
                          Object {
                            "name": "test",
                          },
                        ],
                      }
        }
        store={
                Object {
                        "dispatch": [Function],
                        "getState": [Function],
                        "replaceReducer": [Function],
                        "subscribe": [Function],
                        Symbol(observable): [Function],
                      }
        } />,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": StatelessComponent {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "props": Object {
          "network": Object {
            "nodes": Array [
              Object {
                "name": "test",
              },
            ],
          },
          "store": Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          },
        },
        "refs": Object {},
<<<<<<< HEAD
        "state": null,
=======
        "renderedElement": <Droppable
          acceptsDraggableType={null}
          droppableName={null}
          network={
                    Object {
                              "nodes": Array [
                                Object {
                                  "name": "test",
                                },
                              ],
                            }
          }
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          updateZone={[Function]} />,
        "state": Object {
          "storeState": undefined,
        },
        "stateProps": Object {},
        "statePropsPrecalculationError": null,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
>>>>>>> update snapshots
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
<<<<<<< HEAD
        "_currentElement": <div
          className="scrollable">
          <Connect(Droppable)
                    network={
                              Object {
                                        "nodes": Array [
                                          Object {
                                            "name": "test",
                                          },
                                        ],
                                      }
                    }
                    store={
                              Object {
                                        "dispatch": [Function],
                                        "getState": [Function],
                                        "replaceReducer": [Function],
                                        "subscribe": [Function],
                                        Symbol(observable): [Function],
                                      }
                    } />
</div>,
        "_debugID": 2,
        "_renderedOutput": <div
          className="scrollable">
          <Connect(Droppable)
                    network={
                              Object {
                                        "nodes": Array [
                                          Object {
                                            "name": "test",
                                          },
                                        ],
                                      }
                    }
                    store={
                              Object {
                                        "dispatch": [Function],
                                        "getState": [Function],
                                        "replaceReducer": [Function],
                                        "subscribe": [Function],
                                        Symbol(observable): [Function],
                                      }
                    } />
</div>,
=======
        "_currentElement": <Droppable
          acceptsDraggableType={null}
          droppableName={null}
          network={
                    Object {
                              "nodes": Array [
                                Object {
                                  "name": "test",
                                },
                              ],
                            }
          }
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          updateZone={[Function]} />,
        "_debugID": 2,
        "_renderedOutput": <Droppable
          acceptsDraggableType={null}
          droppableName={null}
          network={
                    Object {
                              "nodes": Array [
                                Object {
                                  "name": "test",
                                },
                              ],
                            }
          }
          store={
                    Object {
                              "dispatch": [Function],
                              "getState": [Function],
                              "replaceReducer": [Function],
                              "subscribe": [Function],
                              Symbol(observable): [Function],
                            }
          }
          updateZone={[Function]} />,
>>>>>>> update snapshots
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <Scrollable
    network={
        Object {
            "nodes": Array [
              Object {
                "name": "test",
              },
            ],
          }
    }
    store={
        Object {
            "dispatch": [Function],
            "getState": [Function],
            "replaceReducer": [Function],
            "subscribe": [Function],
            Symbol(observable): [Function],
          }
    } />,
}
`;
